### ISSUE TITLE
Improve factoring of various reactor implementations of "internal" FD tracking

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3602]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3602|
|Type|enhancement|
|Created|2009-01-07 18:55:24Z|
|Branch|https://github.com/twisted/twisted/tree/internal-fd-tracking-3602|

Various reactors keep track of the thread waker object in various different ways.  They all need to make sure that `removeAll` doesn't actually remove it.  The factoring of this code could be improved.  In particular, it could be made much simpler to add new internal FDs.

This will help [#10460](https://github.com/twisted/twisted/issues/10460).

<details><summary>Searchable metadata</summary>

```
trac-id__3602 3602
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_internal_fd_tracking_3602 branches/internal-fd-tracking-3602
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1231354524000000 1231354524000000
changetime__1232924601000000 1232924601000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25854]) Branching to 'internal-fd-tracking-3602'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

There was some code in /branches/sigchld-791-3 which is helpful for this.  I'm moving it into this branch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

I have a pet peeve about tickets that say they could "improve factoring" or "make stuff better" - better how?  I am guessing that in this case better factoring simply means decreased duplication.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Yes, decreased duplication is the primary goal.

Build results are here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/internal-fd-tracking-3602

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I made one further change because apparently no one has looked at this yet and I felt like it!

I moved some tests out of twisted/test/ into twisted/internet/test/ in r25993.  New build results are in the same place as the old build results, of course.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

 * Some flakes:
```
twisted/test/test_internet.py:8: 'Connection' imported but unused
twisted/test/test_internet.py:25: 'socket' imported but unused
```
 * I'm not sure I see the point of having `_internalReaders` in `ReactorBase` instead of `PosixReactorBase` (as it's not used inside `ReactorBase` at all). This is minor, though.
 * 
```
+        removedReaders = set([
+                r for r in readers if r not in self._internalReaders]) 
```
  This could be written like that I think:
```
+        removedReaders = set(readers) - self._internalReaders
```

That's it, +1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [26117]) remove unused imports; adjust other imports, copyright statement, and module docstring; replace list comprehension with a simpler set operation

refs [#3602](https://github.com/twisted/twisted/issues/3602)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [26118]) Merge internal-fd-tracking-3602

Author: exarkun, itamarst
Reviewer: therve
Fixes: [#3602](https://github.com/twisted/twisted/issues/3602)

Refactor the reactor's internal tracking of file descriptors, particular
with respect to "internal" file descriptors which are created and used by
the reactor itself, not by applications.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
